### PR TITLE
Merge mysql 5.7.35 part1

### PR DESCRIPTION
**Pull request by ldonoso**
<https://jira.percona.com/browse/PS-7810>
<https://ps57.cd.percona.com/job/percona-server-5.7-param/497/> (without `keyring_vault` suite)
<https://ps57.cd.percona.com/job/percona-server-5.7-asan-param/36> (ASAN)

### CODE REVIEW

**Comment by venkatesh-prasad-v** [comment]
One suggestion from my side: Please mention `Part 1` in the commit message so that it becomes easy to look for the exact commit while merging PS to PXC.

```
Implemented PS-7810 (Merge MySQL 5.7.35 Part 1 up to 6642ea3) (conflicts resolved)
```

File: storage/innobase/buf/buf0flu.cc [hunk 1]
```diff
@@ -2729,20 +2729,11 @@ page_cleaner_flush_pages_recommendation(
...
-	ulint	pages_for_lsn =
-		std::min<ulint>(sum_pages_for_lsn, sum_pages_max);
->>>>>>> 6642ea3d6aec50398cda18a28fa64f7082f5f521
 
 	n_pages = PCT_IO(pct_total);
 	if (age < log_get_max_modified_age_async()) {
 		ulint	pages_for_lsn =
-			std::min<ulint>(sum_pages_for_lsn,
-					srv_max_io_capacity * 2);
+			std::min<ulint>(sum_pages_for_lsn, sum_pages_max);
```
Comment (oleksandr-kachan) [inline storage/innobase/buf/buf0flu.cc:2736]: Just to clarify. Where this change came from?

**Review by oleksandr-kachan — COMMENTED** [review]
LGTM

File: cmake/lz4.cmake [hunk 2]
```diff
@@ -52,13 +52,7 @@ MACRO (FIND_SYSTEM_LZ4)
...
   IF (LZ4_INCLUDE_DIR AND LZ4_SYSTEM_LIBRARY)
     SET(SYSTEM_LZ4_FOUND 1)
-<<<<<<< HEAD
-    CHECK_LZ4_VERSION(${PATH_TO_LZ4})
-    SET(LZ4_INCLUDE_DIR ${PATH_TO_LZ4})
-||||||| a9b0c712de3
-    SET(LZ4_INCLUDE_DIR ${PATH_TO_LZ4})
-=======
->>>>>>> 6642ea3d6aec50398cda18a28fa64f7082f5f521
+    CHECK_LZ4_VERSION(${LZ4_INCLUDE_DIR})
```
Comment (percona-ysorokin) [inline cmake/lz4.cmake:55]: It looks like Oracle updated their bundled lz4 library to `1.9.3` whereas Percona's logic (`CHECK_LZ4_VERSION()` macro) still checks for `1.7.1`. I believe we should update `LIBLZ4_VERSION_REQUIRED` in this file to `1.9.3`.

File: storage/innobase/include/mem0mem.ic [hunk 3]
```diff
@@ -597,17 +597,9 @@ mem_strdupl(
 	ulint		len)	/*!< in: length of str, in bytes */
 {
 	char*	s = static_cast<char*>(ut_malloc_nokey(len + 1));
-<<<<<<< HEAD
 	if (!s) return NULL;
 
 	s[len] = 0;
-||||||| a9b0c712de3
```
Comment (percona-ysorokin) [inline storage/innobase/include/mem0mem.ic]: Any reason why this has been resolved in favor of our changes, not upstream?

**Review by percona-ysorokin — CHANGES_REQUESTED** [review]

**Comment by percona-ysorokin** [comment]
@ldonoso There is also very annoying CMake warning coming from PerconaFT
```
CMake Warning (dev) at storage/tokudb/PerconaFT/cmake_modules/TokuFeatureDetection.cmake:135 (set):
  implicitly converting 'INTEGER' to 'STRING' type.
Call Stack (most recent call first):
  storage/tokudb/PerconaFT/CMakeLists.txt:63 (include)
This warning is for project developers.  Use -Wno-dev to suppress it.
```

---
@percona-ysorokin, updating PerconaFT solved the issue, it was added a fix for this

File: storage/innobase/include/mem0mem.ic (hunk 3 above)
Comment (ldonoso) [reply storage/innobase/include/mem0mem.ic]: The reasoning is that our change is the same as theirs but safer.

If `s` is null and `len > 0`, theirs would dereference a `NULL` ptr but not ours.

**Comment by ldonoso** [comment]
> @ldonoso There is also very annoying CMake warning coming from PerconaFT
> 
> ```
> CMake Warning (dev) at storage/tokudb/PerconaFT/cmake_modules/TokuFeatureDetection.cmake:135 (set):
>   implicitly converting 'INTEGER' to 'STRING' type.
> Call Stack (most recent call first):
>   storage/tokudb/PerconaFT/CMakeLists.txt:63 (include)
> This warning is for project developers.  Use -Wno-dev to suppress it.
> ```

**Comment by percona-ysorokin** [comment]
@ldonoso 
Ubuntu 20.04 Focal (gcc-11, both Debug and RelWithDebInfo)
```
[ 50%] innodb.innodb-multiple-tablespaces       [ fail ]  Found warnings/errors in server log file!
        Test ended at 2021-08-02 20:35:44
line
2021-08-02T20:35:32.484649Z 0 [ERROR] InnoDB: Operating system error number 22 in a file operation.
2021-08-02T20:35:32.484654Z 0 [ERROR] InnoDB: Error number 22 means 'Invalid argument'
2021-08-02T20:35:32.758282Z 2 [ERROR] InnoDB: Operating system error number 22 in a file operation.
2021-08-02T20:35:32.758287Z 2 [ERROR] InnoDB: Error number 22 means 'Invalid argument'
2021-08-02T20:35:36.109233Z 0 [ERROR] InnoDB: Operating system error number 22 in a file operation.
2021-08-02T20:35:36.109239Z 0 [ERROR] InnoDB: Error number 22 means 'Invalid argument'
2021-08-02T20:35:36.448591Z 2 [ERROR] InnoDB: Operating system error number 22 in a file operation.
2021-08-02T20:35:36.448597Z 2 [ERROR] InnoDB: Error number 22 means 'Invalid argument'
^ Found warnings in /home/yura/ws/percona-build-5.7-debug_gcc11/mysql-test/var/log/mysqld.1.err
ok
```

```
[ 50%] innodb.alter_kill                        [ fail ]  Found warnings/errors in server log file!
        Test ended at 2021-08-02 20:38:35
line
2021-08-02T20:38:19.972986Z 0 [ERROR] InnoDB: Error number 22 means 'Invalid argument'
2021-08-02T20:38:20.238097Z 2 [ERROR] InnoDB: Error number 22 means 'Invalid argument'
^ Found warnings in /home/yura/ws/percona-build-5.7-debug_gcc11/mysql-test/var/log/mysqld.1.err
ok
```

```
worker[1] Using MTR_BUILD_THREAD 300, with reserved ports 13000..13009
[ 50%] data_masking.error_not_loaded_module     [ fail ]
        Test ended at 2021-08-02 20:39:37

CURRENT_TEST: data_masking.error_not_loaded_module
mysqltest: At line 1: query 'CREATE FUNCTION gen_blacklist RETURNS STRING SONAME 'data_masking.so'' failed: 1126: Can't open shared library 'data_masking.so' (errno: 2 /home/yura/ws/percona-server/lib/mysql/plugin/data_masking.so: cannot open shared object file: No such file or directory)
safe_process[1591784]: Child process: 1591785, exit: 1
```

File: storage/innobase/include/mem0mem.ic (hunk 3 above)
Comment (percona-ysorokin) [reply storage/innobase/include/mem0mem.ic]: OK, indeed looks safer this way

**Comment by percona-ysorokin** [comment]
@ldonoso One more issue detected by ASan
```
[ 20%] main.bug12969156                         w9 [ fail ]
        Test ended at 2021-08-04 13:09:29

CURRENT_TEST: main.bug12969156
mysqltest: At line 16: query 'UNINSTALL PLUGIN daemon_example' failed: 2013: Lost connection to MySQL server during query
```

```
==90890==ERROR: AddressSanitizer: stack-buffer-overflow on address 0x7fe746d06d14 at pc 0x7fe760f5b017 bp 0x7fe746d06cd0 sp 0x7fe746d06478
WRITE of size 24 at 0x7fe746d06d14 thread T16777215
    #0 0x7fe760f5b016 in __interceptor_sigaltstack ../../../../src/libsanitizer/sanitizer_common/sanitizer_common_interceptors.inc:9996
    #1 0x7fe760fd9bfd in __sanitizer::UnsetAlternateSignalStack() ../../../../src/libsanitizer/sanitizer_common/sanitizer_posix_libcdep.cpp:194
    #2 0x7fe760fc99ec in __asan::AsanThread::Destroy() ../../../../src/libsanitizer/asan/asan_thread.cpp:104
    #3 0x7fe760ab4430 in __nptl_deallocate_tsd nptl/pthread_create.c:303
    #4 0x7fe760ab5470 in __nptl_deallocate_tsd nptl/pthread_create.c:258
    #5 0x7fe760ab5470 in start_thread nptl/pthread_create.c:484
    #6 0x7fe760655d52 in clone (/lib/x86_64-linux-gnu/libc.so.6+0x117d52)

Address 0x7fe746d06d14 is located in stack of thread T26 at offset 340 in frame
    #0 0x7fe746d0a55c in mysql_heartbeat(void*) /home/yura/ws/percona-server/plugin/daemon_example/daemon_example.cc:62

  This frame has 4 object(s):
    [48, 56) 'result' (line 66)
    [80, 112) '_db_stack_frame_' (line 63)
    [144, 200) 'tm_tmp' (line 67)
    [240, 340) 'buffer' (line 65) <== Memory access at offset 340 overflows this variable
HINT: this may be a false positive if your program uses some custom stack unwind mechanism, swapcontext or vfork
      (longjmp and C++ exceptions *are* supported)
Thread T26 created by T25 here:
    #0 0x7fe760f5f6d5 in __interceptor_pthread_create ../../../../src/libsanitizer/asan/asan_interceptors.cpp:216
    #1 0x557ccbbcb857 in my_thread_create /home/yura/ws/percona-server/mysys/my_thread.c:104
    #2 0x7fe746d0b21a in daemon_example_plugin_init /home/yura/ws/percona-server/plugin/daemon_example/daemon_example.cc:148
    #3 0x557ccb4c69c7 in plugin_initialize /home/yura/ws/percona-server/sql/sql_plugin.cc:1279
    #4 0x557ccb4d19cd in mysql_install_plugin /home/yura/ws/percona-server/sql/sql_plugin.cc:2279
    #5 0x557ccb4d218f in Sql_cmd_install_plugin::execute(THD*) /home/yura/ws/percona-server/sql/sql_plugin.cc:4664
    #6 0x557ccb47695e in mysql_execute_command(THD*, bool) /home/yura/ws/percona-server/sql/sql_parse.cc:5160
    #7 0x557ccb47977c in mysql_parse(THD*, Parser_state*, bool) /home/yura/ws/percona-server/sql/sql_parse.cc:5952
    #8 0x557ccb47b6c2 in dispatch_command(THD*, COM_DATA const*, enum_server_command) /home/yura/ws/percona-server/sql/sql_parse.cc:1544
    #9 0x557ccb47de1d in do_command(THD*) /home/yura/ws/percona-server/sql/sql_parse.cc:1065
    #10 0x557ccb6ac294 in handle_connection /home/yura/ws/percona-server/sql/conn_handler/connection_handler_per_thread.cc:325
    #11 0x557ccbbfabb0 in pfs_spawn_thread /home/yura/ws/percona-server/storage/perfschema/pfs.cc:2198
    #12 0x7fe760ab544f in start_thread nptl/pthread_create.c:473
```

**Comment by percona-ysorokin** [comment]
In addition, please add `leak:/usr/bin/python3.9` suppression to `mysql-test/lsan.supp`.
Also, `--source include/have_grep.inc` does not seem to work under ASan casing a few false positives
```
[  5%] main.explain_for_connection_small_json   w12 [ skipped ]  Test requires gnu grep
[  6%] main.explain_for_connection_small_trad   w12 [ skipped ]  Test requires gnu grep
```
It looks like the `grep` utility has some leaks and always return `42` under ASan.

**Comment by percona-ysorokin** [comment]
One more issue.
This is probably optional as it is not connected with the current merge but would be really nice to fix.
The following 2 tests hangs forever under ASan
```
main.mysqld_safe
main.percona_processlist_tid
```
Most probably this is also connected with exit code 42 returned by executables in case of leaks under ASan.

**Comment by ldonoso** [comment]
@percona-ysorokin, all the issues are fixable, as they are not related with the merge, I have created some separate tickets to solve them and to fix them in 8.0 as well:
<https://jira.percona.com/browse/PS-7841>
<https://jira.percona.com/browse/PS-7842>
<https://jira.percona.com/browse/PS-7843>
<https://jira.percona.com/browse/PS-7844>

Please, could you approve this PR if you think it is fine?

**Review by percona-ysorokin — APPROVED** [review]
LGTM

**Pull request merged**